core/tee/tee_rpmb_fs.c

```
@@ -1472,54 +1514,364 @@ static TEE_Result tee_rpmb_get_max_block(uint16_t dev_id, uint32_t *max_block)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint32_t fat_entry_buf_idx;
+	/* Use a temp var to avoid compiler warning if caching disabled. */
+	uint32_t max_cache_entries = CFG_RPMB_FS_CACHE_SIZE;
+
+	/* Nothing to update if the cache is not initialized. */
+	if (!fat_entry_dir)
+		goto out;
+
+	/* Sanity check for offset of FAT address. */
+	if ((fat_address - RPMB_FS_FAT_START_ADDRESS) %
+	     sizeof(struct rpmb_fat_entry)) {
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto out;
+	}
+
+	/* Determine index in cache. */
+	fat_entry_buf_idx = (fat_address - RPMB_FS_FAT_START_ADDRESS) /
+			     sizeof(struct rpmb_fat_entry);
+
+	/* Only need to write if index points to an entry in cache. */
+	if (fat_entry_buf_idx < max_cache_entries) {
+		memcpy(fat_entry_dir->rpmb_fat_entry_buf + fat_entry_buf_idx,
+		       fat_entry, sizeof(struct rpmb_fat_entry));
+	}
+
+out:
+	return res;
+}
+
+/**
+ * fat_entry_dir_get_next: Get next FAT FS entry.
+ * Read either from cache/buffer, or by reading from RPMB storage if the
+ * elements in the buffer/cache are fully read. When reading in from RPMB
+ * storage, the buffer is overwritten in case caching is disabled.
+ * In case caching is enabled, the cache is either further filled, or a
+ * temporary buffer populated if the cache is already full.
+ * The FAT FS entry is written to fat_entry. The respective address in RPMB
+ * storage is written to fat_address, if not NULL. When the last FAT FS entry
+ * was previously read, the function indicates this case by writing a NULL
+ * pointer to fat_entry.
+ * Returns a value different TEE_SUCCESS if the next FAT FS entry could not be
+ * retrieved.
+ */
+static TEE_Result fat_entry_dir_get_next(struct rpmb_fat_entry **fat_entry,
+					 uint32_t *fat_address)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct rpmb_fat_entry *fe = NULL;
+	uint32_t num_elems_read;
+	uint32_t fat_address_local;
+
+	/* Must call fat_entry_dir_init first. */
+	if (!fat_entry_dir || !fat_entry)
+		goto out;
```

Since a NULL value indicates a programming error with the caller (not a runtime error), I think I'd rather let the function crash. Opinions?

---

Agreed. How exactly would this change the code? Right now we're returning TEE_ERROR_GENERIC in this case. Did you think of an assert?

---

Yes